pkg/controller/v1alpha2/core/scopes/healthscope/healthscope_controller.go

```diff
@@ -152,86 +168,98 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 
 	log = log.WithValues("uid", hs.GetUID(), "version", hs.GetResourceVersion())
 
-	hsStatus := r.GetScopeHealthStatus(ctx, hs)
+	scopeCondition, wlConditions := r.GetScopeHealthStatus(ctx, hs)
 	log.Debug("Successfully ran health check", "scope", hs.Name)
 	r.record.Event(hs, event.Normal(reasonHealthCheck, "Successfully ran health check"))
 
 	elapsed := time.Since(start)
-
-	if hsStatus.IsHealthy {
-		hs.Status.Health = "healthy"
-	} else {
-		hs.Status.Health = "unhealthy"
-	}
-
-	diagnosis, err := json.Marshal(hsStatus)
-	if err != nil {
-		return reconcile.Result{}, errors.Wrap(err, errMarshalScopeDiagnosiis)
-	}
-
-	// sava diagnosis into status.conditions
-	//TODO(roywang) is there a better way to show diagnosis
-	hs.SetConditions(v1alpha1.ReconcileSuccess().
-		WithMessage(string(diagnosis)))
+	hs.Status.ScopeHealthCondition = scopeCondition
+	hs.Status.HealthConditions = wlConditions
 
 	return reconcile.Result{RequeueAfter: interval - elapsed}, errors.Wrap(r.client.Status().Update(ctx, hs), errUpdateHealthScopeStatus)
 }
 
 // GetScopeHealthStatus get the status of the healthscope based on workload resources.
-func (r *Reconciler) GetScopeHealthStatus(ctx context.Context, healthScope *v1alpha2.HealthScope) HealthCondition {
-	hsStatus := HealthCondition{
-		Target: runtimev1alpha1.TypedReference{
+func (r *Reconciler) GetScopeHealthStatus(ctx context.Context, healthScope *v1alpha2.HealthScope) (HealthCondition, []*HealthCondition) {
+	log := r.log.WithValues("get scope health status", healthScope.GetName())
+	scopeCondition := HealthCondition{
+		TargetWorkload: runtimev1alpha1.TypedReference{
 			APIVersion: healthScope.APIVersion,
 			Kind:       healthScope.Kind,
 			Name:       healthScope.Name,
 			UID:        healthScope.UID,
 		},
-		IsHealthy:     true, //if no workload referenced, scope is healthy by default
-		SubConditions: []*HealthCondition{},
+		HealthStatus: StatusHealthy, //if no workload referenced, scope is healthy by default
+	}
+	scopeWLRefs := healthScope.Spec.WorkloadReferences
+	if len(scopeWLRefs) == 0 {
+		return scopeCondition, []*HealthCondition{}
 	}
+
 	timeout := defaultTimeout
 	if healthScope.Spec.ProbeTimeout != nil {
 		timeout = time.Duration(*healthScope.Spec.ProbeTimeout) * time.Second
 	}
 	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 
-	scopeWLRefs := healthScope.Spec.WorkloadReferences
 	// process workloads concurrently
-	wlStatusCs := make(chan *HealthCondition, len(scopeWLRefs))
+	workloadHealthConditionsC := make(chan *HealthCondition, len(scopeWLRefs))
 	var wg sync.WaitGroup
 	wg.Add(len(scopeWLRefs))
 
 	for _, workloadRef := range scopeWLRefs {
 		go func(resRef runtimev1alpha1.TypedReference) {
 			defer wg.Done()
-			var wlStatusC *HealthCondition
+			var wlHealthCondition *HealthCondition
+
+			wlHealthCondition = r.traitChecker.Check(ctx, r.client, resRef, healthScope.GetNamespace())
```

Comment:
why traitChecker has higher priority than built-in ones？

Reply:
I think it's possible for user to override health check logic on these "built-in" workloads.